toucan_connectors/common.py

```diff
@@ -10,25 +13,125 @@ def nosql_apply_parameters_to_query(query, parameters):
     Instead use your client library parameter substitution method.
     https://www.owasp.org/index.php/Query_Parameterization_Cheat_Sheet
     """
+    def _prepare_parameters(p):
+        if isinstance(p, str):
+            return f'"{p}"'
+        elif isinstance(p, list):
+            return [_prepare_parameters(e) for e in p]
+        elif isinstance(p, dict):
+            return {k: _prepare_parameters(v) for k, v in p.items()}
+        else:
+            return p
+
+    def _prepare_result(res):
+        if isinstance(res, str):
+            return ast.literal_eval(res)
+        elif isinstance(res, list):
+            return [_prepare_result(e) for e in res]
+        elif isinstance(res, dict):
+            return {k: _prepare_result(v) for k, v in res.items()}
+        else:
+            return res
+
+    def _render_query(query, parameters):
+        if isinstance(query, dict):
+            return {key: _render_query(value, parameters)
+                    for key, value in deepcopy(query).items()}
+        elif isinstance(query, list):
+            return [_render_query(elt, parameters) for elt in deepcopy(query)]
+        elif type(query) is str:
+            clean_p = deepcopy(parameters)
+            # Add quotes to string parameters to keep type if not complex
+            if (re.compile(r"^{{[^({{)}]*}}$").match(query)
```

Comment:
maybe compile the regexes in a CONSTANT to make it more readable and probably gain a bit of performance